### ISSUE TITLE
RISC-V: error: 2-byte data relocations not supported

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1143]
From CI: https://travis-ci.com/github/ClangBuiltLinux/continuous-integration/jobs/377856754

```
<instantiation>:8:9: error: 2-byte data relocations not supported
 .short section_count
        ^
<instantiation>:51:8: error: expected relocatable expression
 .long (section_table - .) / 8
       ^
make[4]: *** [scripts/Makefile.build:364: arch/riscv/kernel/head.o] Error 1
```

Presumably caused by https://git.kernel.org/riscv/c/07f31baa5943256c224a32df82c503f49572c057.

**Comment by ihalip** [comment]
Oh boy, integrated assembler [doesn't support](https://github.com/llvm/llvm-project/blob/ae1396c7d4d83366695137f69f046719fd199408/llvm/lib/MC/MCExpr.cpp#L877) the `/` operand in such an expression.

**Comment by jrtc27** [comment]
This is related to linker relaxation (note that -mno-relax also makes it work). It seems LLVM isn't quite smart enough to know it can calculate those relocations even though they're in a section that's already been tainted with relaxable code.

**Comment by nathanchance** [comment]
@jrtc27 with this diff:

```diff
diff --git a/arch/riscv/Makefile b/arch/riscv/Makefile
index 8c29e553ef7f..2531f8bbc484 100644
--- a/arch/riscv/Makefile
+++ b/arch/riscv/Makefile
@@ -57,7 +57,14 @@ ifeq ($(CONFIG_PERF_EVENTS),y)
         KBUILD_CFLAGS += -fno-omit-frame-pointer
 endif

-KBUILD_CFLAGS_MODULE += $(call cc-option,-mno-relax)
+mno-relax := $(call cc-option,-mno-relax)
+
+ifeq ($(CONFIG_LD_IS_LLD),y)
+KBUILD_CFLAGS += $(mno-relax)
+KBUILD_AFLAGS += $(mno-relax)
+endif
+
+KBUILD_CFLAGS_MODULE += $(mno-relax)

 # GCC versions that support the "-mstrict-align" option default to allowing
 # unaligned accesses.  While unaligned accesses are explicitly allowed in the
```

we get farther but we are still broken with the integrated assembler:

```
In file included from init/main.c:17:
In file included from ./include/linux/module.h:18:
In file included from ./include/linux/elf.h:6:
In file included from ./arch/riscv/include/asm/elf.h:14:
In file included from ./arch/riscv/include/asm/cacheinfo.h:9:
In file included from ./include/linux/cacheinfo.h:6:
In file included from ./include/linux/cpu.h:17:
In file included from ./include/linux/node.h:18:
In file included from ./include/linux/device.h:16:
In file included from ./include/linux/energy_model.h:6:
In file included from ./include/linux/jump_label.h:117:
./arch/riscv/include/asm/jump_label.h:28:36: error: Unsupported relocation type
                "       .long           1b - ., %l[label] - .   \n\t"
                                                                  ^
<inline asm>:9:10: note: instantiated into assembly here
                .dword  init_on_alloc - .
                        ^
In file included from init/main.c:17:
In file included from ./include/linux/module.h:18:
In file included from ./include/linux/elf.h:6:
In file included from ./arch/riscv/include/asm/elf.h:14:
In file included from ./arch/riscv/include/asm/cacheinfo.h:9:
In file included from ./include/linux/cacheinfo.h:6:
In file included from ./include/linux/cpu.h:17:
In file included from ./include/linux/node.h:18:
In file included from ./include/linux/device.h:16:
In file included from ./include/linux/energy_model.h:6:
In file included from ./include/linux/jump_label.h:117:
./arch/riscv/include/asm/jump_label.h:28:36: error: Unsupported relocation type
                "       .long           1b - ., %l[label] - .   \n\t"
                                                                  ^
<inline asm>:9:10: note: instantiated into assembly here
                .dword  init_on_free - .
                        ^
In file included from init/main.c:17:
In file included from ./include/linux/module.h:18:
In file included from ./include/linux/elf.h:6:
In file included from ./arch/riscv/include/asm/elf.h:14:
In file included from ./arch/riscv/include/asm/cacheinfo.h:9:
In file included from ./include/linux/cacheinfo.h:6:
In file included from ./include/linux/cpu.h:17:
In file included from ./include/linux/node.h:18:
In file included from ./include/linux/device.h:16:
In file included from ./include/linux/energy_model.h:6:
In file included from ./include/linux/jump_label.h:117:
./arch/riscv/include/asm/jump_label.h:28:36: error: Unsupported relocation type
                "       .long           1b - ., %l[label] - .   \n\t"
                                                                  ^
<inline asm>:9:10: note: instantiated into assembly here
                .dword  init_on_free - .
                        ^
3 errors generated.
```

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/riscv/include/asm/jump_label.h

**Comment by jrtc27** [comment]
Hmm. GNU as will always use ADD32/64+SUB32/64 to do symbol differences. LLVM will use that for -mrelax, but will optimise it to 32_PCREL for `.long` and fail to find a suitable relocation for `.quad`. The semantics of the two aren't quite the same though when it comes to symbol preemptibility.

**Comment by jrtc27** [comment]
Is there a reason you need a 64-bit integer? That seems like a bit of a waste when the code model only supports 2 GiB, and using a 32-bit integer would also work around the problem.

**Comment by nickdesaulniers** [comment]
The inline asm goto in jump_label.h is encoding `struct jump_entry`'s in the `__jump_table` section.  Those are defined as:

```c
// include/linux/jump_label.h
122 struct jump_entry {                                                                
123   s32 code;                                                                        
124   s32 target;                                                                      
125   long key; // key may be far away from the core kernel under KASLR                
126 };
```
which is the same definition for all architectures supported by the Linux kernel. That code in particular [is already insanely complex](http://nickdesaulniers.github.io/blog/2020/04/06/off-by-two/); I don't see it changing to support one arch.

**Comment by nickdesaulniers** [comment]
More thoughts posted in https://github.com/ClangBuiltLinux/linux/issues/1020#issuecomment-828689118, but right now, it seems like there's no path forward today; if we enable `-mno-relax`, integrated assembler support regresses in other ways.

**Comment by nickdesaulniers** [comment]
Here's a reduced test case for the specific error:
```sh
$ cat clang --target=riscv64-linux-gnu -c head.s
 .macro __EFI_PE_HEADER
 .short section_count
 .endm
 __EFI_PE_HEADER
$ clang --target=riscv64-linux-gnu -c head.s           
<instantiation>:1:8: error: 2-byte data relocations not supported
.short section_count
       ^
```
https://llvm.org/pr50156

**Comment by MaskRay** [comment]
GNU as doesn't support this either, even in -mno-relax mode.
```
% cat a.s
section_count:
.short section_count
% riscv64-linux-gnu-as -mno-relax a.s -o a.o
a.s: Assembler messages:
a.s:2: Error: cannot represent BFD_RELOC_16 relocation in object file
```

x86/aarch64/etc support this, either local (resolved) or non-local (2-byte absolute relocation).

**Comment by nickdesaulniers** [comment]
`-mno-relax` landed in https://git.kernel.org/pub/scm/linux/kernel/git/riscv/linux.git/commit/?h=fixes&id=ec3a5cb61146c91f0f7dcec8b7e7157a4879a9ee (#1020). https://reviews.llvm.org/D103539 resolves #1023. https://reviews.llvm.org/D103539 does not resolve the error from [@nathanchance 's comment above](https://github.com/ClangBuiltLinux/linux/issues/1143#issuecomment-770125754).
```
./arch/riscv/include/asm/jump_label.h:28:36: error: Unsupported relocation type
                "       .long           1b - ., %l[label] - .   \n\t"
                                                                  ^
```

**Comment by compnerd** [comment]
I need to fix another case in the changes in [D103539](https://reviews.llvm.org/D103539), but noticed another minor thing that we don't handle properly.  This is next on my list of things to address.

**Comment by nickdesaulniers** [comment]
big thanks @compnerd !
https://reviews.llvm.org/rGbbea64250f65480d787e1c5ff45c4de3ec2dcda8
